### PR TITLE
Fix issues with styles in addon directory #56

### PR DESCRIPTION
**Pull request by dukex**
I don't know if this approach is the better way to fix the issues, maybe I need understand more about broccoli and ember-cli internals functions, but It's fix the problems(in ember-cli-sass) mentioned on https://github.com/ember-cli/ember-cli/issues/4084 and fix issue https://github.com/aexmachina/ember-cli-sass/issues/56

For understand why I made this checks please see @stefanpenner anwser on https://github.com/ember-cli/ember-cli/issues/4258

### CODE REVIEW

**Comment by stefanpenner** [comment]
i'll try to make time later this week to review this.

File: index.js [hunk 1]
```diff
@@ -5,6 +5,10 @@ var mergeTrees = require('broccoli-merge-trees');
 var merge = require('merge');
 var fs = require('fs');
 
+function isAddonTree(inputPath) {
+  return inputPath === "/";
+}
+
```
Comment (aexmachina) [inline index.js]: This function should be at the bottom of the file.

File: index.js [hunk 2]
```diff
@@ -27,13 +31,19 @@ SASSPlugin.prototype.toTree = function(tree, inputPath, outputPath, inputOptions
     }
     var output = options.outputPaths[file];
     if (input) {
+      // the includePaths on addon point to addon/styles, input should be relative to includePaths
+      if(isAddonTree(inputPath)) { input = input.replace(tree + "/", '') }
```
Comment (aexmachina) [inline index.js]:
```js
if (isAddonTree(inputPath)) {
  input = input.replace(tree + '/', '')
}
```

File: index.js [hunk 3]
```diff
@@ -27,13 +31,19 @@ SASSPlugin.prototype.toTree = function(tree, inputPath, outputPath, inputOptions
...
     }
     return trees;
   }, []);
 
   function tryFile(file) {
-    var filePath = path.join('.', inputPath, file);
+    var root;
+    // When ember-cli preprocess addon style the inputPath is '/',
+    // add the '/addon/styles' to try found addon sass files
+    if(isAddonTree(inputPath)) { root = tree } else { root = "." }
```
Comment (aexmachina) [inline index.js]:
```js
root = isAddonTree(inputPath) ? tree : '.';
```

**Comment by aexmachina** [comment]
I'm okay with this solution, but only if it relies on stable behaviour from ember-cli. What do you think @stefanpenner?

File: index.js (hunk 3 above)
Comment (dukex) [inline index.js]: Ohhh, my bad, I'm working with coffeescript a long time I forget about ternary

**Comment by dukex** [comment]
@stefanpenner any feedback?

**Comment by stefanpenner** [comment]
> @stefanpenner any feedback?

haven't had a chance yet

File: index.js [hunk 4]
```diff
@@ -27,13 +27,20 @@ SASSPlugin.prototype.toTree = function(tree, inputPath, outputPath, inputOptions
...
     }
     return trees;
   }, []);
 
   function tryFile(file) {
-    var filePath = path.join('.', inputPath, file);
+    // When ember-cli preprocess addon style the inputPath is '/',
+    // add the '/addon/styles' to try found addon sass files
+    var root = isAddonTree(inputPath) ? tree : ".";
+    var filePath = path.join(root, inputPath, file);
```
Comment (justin-lau) [inline index.js]: @dukex this line is throwing an error when I try to build an application that uses an addon with this branch:

```
TypeError: Arguments to path.join must be strings
    at Object.posix.join (path.js:488:13)
    at tryFile (node_modules/ember-cli-sass/index.js:48:25)
```

where `root` evaluated to `{ read: [Function], cleanup: [Function] }`

Comment (justin-lau) [inline index.js]: I have submitted a PR dukex/ember-cli-sass#1 to fix the mentioned issue.

**Comment by dukex** [comment]
Thanks @justin-lau!

**Comment by ahouchens** [comment]
Verified that this pull request works perfectly with ember-cli for preprocessing SASS. :) Thanks so much.

**Comment by begedin** [comment]
Is there anything the consuming app needs to be doing in order for this PR to work?

It works from the dummy app provided by a developing addon, but vendor.css is empty if I consume the developing addon through an external app.

To elaborate:

* Addon `some-random-addon` is using `ember-cli-sass` (this branch)
  * Only `.scss` file I have in the addon is `addon/styles/addon.scss`
* App `my-app` is using `some-random-addon`.
* sass works in dummy app provided by `some-random-addon` - `vendor.css` has content
* sass doesn't work in `my-app` - `vendor.css` is empty.

* Does `my-app` need to depend on `ember-cli-sass` as well?
* Does `my-app` need an empty `app.css|.scss`?
* Does `some-random-addon` need an empty `app/styles/app.css|.scss`?

**Comment by justin-lau** [comment]
Below is what works for me:

1. `some-random-addon` should place `ember-cli-sass` under `dependencies` instead of `devDependencies`.

2. If you need Sass in `my-app`, it should also use this same branch of `emebr-cli-sass`, but it is okay to place it under `devDependencies`.

**Comment by begedin** [comment]
I'm sorry. It was a common mistake on my part. I used `npm link` to test out the addon and forgot to add it to the list of app's dependencies.

Just in case someone repeats my mistake, the full list of steps to get it working:

* initialize a new addon

```
ember addon some-addon
```

* add this branch as a dependency (**not devDependency**):

```
cd some-addon
npm install --save https://github.com/dukex/ember-cli-sass.git#bugfix/addon-styles
```

* modify the addon's `index.js`

```JavaScript
// index.js for 'some-addon'
module.exports = {
  name: 'some-addon',
  included: function(app) {
    this._super.included(app);
  }
}
```

* create `addon.scss`

```
mkdir some-addon/addon/styles
touch some-addon/addon/styles/addon.scss
```

* add whatever it is you want to `addon.scss`.
* create a new app
```
cd path/to/parent
ember new my-app
```
* use `npm link` to link the addon with the app
```
cd path/to/some-addon
npm link
cd path/to/my-app
npm link some-addon
```

* add the addon as a dev dependency for the app, manually, by editing `package.json` of the app
```JSON
"devDependencies": {
  ...
  "some-addon": "*"
}
```
* run the app, it should work now

**Comment by ahouchens** [comment]
Perfect walkthrough @begedin.

**Comment by Fed03** [comment]
+1

**Comment by dukex** [comment]
@justin-lau found I bug in windows, I'll check @justin-lau's pull request and remove the '[WIP]' prefix

**Comment by vitch** [comment]
I can confirm that the code in this PR works for me as well in my addon.

If you want the code to work whilst developing the addon (e.g. running `ember s` in the addon directory to run the `tests/dummy` application) you will also need to add the `extension` in the `sassOptions` when initialising the addon in `ember-cli-build.js` as mentioned in #56 e.g.

```js
var app = new EmberApp({
  sassOptions: {
    extension: 'scss'
  }
});
```

**Comment by dukex** [comment]
@vitch do you got same error without `extension` options?

**Comment by vitch** [comment]
Without the extension option when compiling the addons dummy app with this branch I get:

```
Arguments to path.resolve must be strings
TypeError: Arguments to path.resolve must be strings
    at Object.posix.resolve (path.js:439:13)
    at Object.posix.relative (path.js:505:16)
    at tryFile (/path/to/project/node_modules/ember-cli-sass/index.js:57:17)
    at /path/to/project/node_modules/ember-cli-sass/index.js:26:15
    at Array.reduce (native)
    at SASSPlugin.toTree (/path/to/project/node_modules/ember-cli-sass/index.js:20:48)
    at /path/to/project/node_modules/ember-component-css/node_modules/ember-cli-preprocess-registry/preprocessors.js:186:26
    at Array.forEach (native)
    at processPlugins (/path/to/project/node_modules/ember-component-css/node_modules/ember-cli-preprocess-registry/preprocessors.js:184:11)
    at module.exports.preprocessCss (/path/to/project/node_modules/ember-component-css/node_modules/ember-cli-preprocess-registry/preprocessors.js:154:10)
```

**Comment by dukex** [comment]
@vitch I will investigate it
/cc @justin-lau

**Comment by ahouchens** [comment]
I only get that error if I did not run `npm install` or `bower install`.

**Comment by sandstrom** [comment]
@dukex If this issue is resolved, will that also solve https://github.com/ember-cli/ember-cli/issues/2905?

**Comment by dukex** [comment]
@vitch I fix it! Update the repository and check again

Sorry delay

**Comment by dukex** [comment]
@sandstrom I believe not, the ember-cli/ember-cli#2905 example app don't uses ember-cli-sass and this pull request don't has effect in that example

**Comment by aexmachina** [comment]
I've gotta say that I think this is really an issue with ember-cli, and it looks that ember-cli/ember-cli#3344 fixes the problem in ember-cli. This PR seems to introduce quite a bit of coupling to ember-cli internals, so I'd prefer to wait until the underlying issue is resolved in ember-cli.

In the meantime addon authors can simply specify the `extension` option as a workaround.

**Comment by ahouchens** [comment]
@aexmachina "In the meantime addon authors can simply specify the extension option as a workaround."

That's not accurate. That is another issue which this pull request doesn't address. The only way my files get preprocessed is if I specify the `extension` option AND point to this PR in my package.json.

**Comment by aexmachina** [comment]
Aha, okay thanks for the info. I will reopen and review tomorrow.

**Comment by ahouchens** [comment]
Great thanks @aexmachina!

File: index.js [hunk 5]
```diff
@@ -33,8 +33,36 @@ SASSPlugin.prototype.toTree = function(tree, inputPath, outputPath, inputOptions
   }, []);
 
   function tryFile(file) {
-    var filePath = path.join('.', inputPath, file);
-    return fs.existsSync(filePath) ? filePath : false;
+    /**
```
Comment (aexmachina) [inline index.js]: Double-asterisk `**` denotes JSDoc, so this should be a single asterisk.

File: index.js [hunk 6]
```diff
@@ -33,8 +33,36 @@ SASSPlugin.prototype.toTree = function(tree, inputPath, outputPath, inputOptions
...
+     * When building an hosting app that indirectly uses `ember-cli-sass`
+     * through an addon, `tree` is an object with a `read()` function that
+     * returns `"/path/to/your/addon-name/addon/styles"`.
+     */
+    var treeString = tree.read ? tree.read() : tree;
+
+    /**
+     * When tree is a 'TreeMerger (stylesAndVendor)'
+     * the 'read' returns a object, so use the a relative treeString(.)
+     */
```
Comment (aexmachina) [inline index.js]: Double-slash comments would be better here `//`

**Comment by aexmachina** [comment]
@stefanpenner any feedback on this? It seems quite tied to implementation details in ember-cli, but I'm happy to go with your advice on this.

**Comment by dukex** [comment]
I believe the https://github.com/ember-cli/ember-cli/pull/3344 will help us to remove this hack

**Comment by aexmachina** [comment]
I've asked @rwjblue for an opinion on whether we should add this hack or wait for the underlying issues to be fixed in ember-cli.

**Comment by dukex** [comment]
I tried to run https://github.com/ember-cli/ember-cli/pull/3344 but many tests are failed

**Comment by alexismangin** [comment]
Hi there, when will this pull request supposed be merged into master?

**Comment by aexmachina** [comment]
We're trying to avoid merging this, because it's basically a hack to get around a problem in ember-cli which should be fixed when ember-cli/ember-cli#3344 is merged. @stefanpenner should we just merge this and then pull it out when the problem in ember-cli is resolved?

**Comment by stefanpenner** [comment]
cc @rwjblue whats the status if your fix?

**Comment by dukex** [comment]
Hi everyone, I made pull request #78 to fix the issue addressed in this pull request, but without hack and using other approach, now I using the #78 in my projects and works fine.

**Comment by dukex** [comment]
Closing in favor of #78

**Pull request closed without merging**